src/Protocol/PortableContact.php

```diff
@@ -0,0 +1,1600 @@
+			&& (normalise_link($profile) == normalise_link($data["alias"]))
+			&& (normalise_link($profile) != normalise_link($data["url"]))
+		) {
+			// Delete the old entry
+			q("DELETE FROM `gcontact` WHERE `nurl` = '%s'", dbesc(normalise_link($profile)));
+			q("DELETE FROM `glink` WHERE `gcid` = %d", intval($gcontacts[0]["id"]));
+	
+			$gcontact = array_merge($gcontacts[0], $data);
+	
+			$gcontact["server_url"] = $data["baseurl"];
+	
+			try {
+				$gcontact = GContact::sanitize($gcontact);
+				GContact::update($gcontact);
+	
+				self::lastUpdated($data["url"], $force);
+			} catch (Exception $e) {
+				logger($e->getMessage(), LOGGER_DEBUG);
+			}
+	
+			logger("Profile ".$profile." was deleted", LOGGER_DEBUG);
+			return false;
+		}
+	
+		if (($data["poll"] == "") || (in_array($data["network"], array(NETWORK_FEED, NETWORK_PHANTOM)))) {
+			q(
+				"UPDATE `gcontact` SET `last_failure` = '%s' WHERE `nurl` = '%s'",
+				dbesc(datetime_convert()),
+				dbesc(normalise_link($profile))
+			);
+	
+			logger("Profile ".$profile." wasn't reachable (profile)", LOGGER_DEBUG);
+			return false;
+		}
+	
+		$contact = array_merge($contact, $data);
+	
+		$contact["server_url"] = $data["baseurl"];
+	
+		GContact::update($contact);
+	
+		$feedret = z_fetch_url($data["poll"]);
+	
+		if (!$feedret["success"]) {
+			q(
+				"UPDATE `gcontact` SET `last_failure` = '%s' WHERE `nurl` = '%s'",
+				dbesc(datetime_convert()),
+				dbesc(normalise_link($profile))
+			);
+	
+			logger("Profile ".$profile." wasn't reachable (no feed)", LOGGER_DEBUG);
+			return false;
+		}
+	
+		$doc = new DOMDocument();
+		@$doc->loadXML($feedret["body"]);
+	
+		$xpath = new DomXPath($doc);
+		$xpath->registerNamespace('atom', "http://www.w3.org/2005/Atom");
+	
+		$entries = $xpath->query('/atom:feed/atom:entry');
+	
+		$last_updated = "";
+	
+		foreach ($entries as $entry) {
+			$published = $xpath->query('atom:published/text()', $entry)->item(0)->nodeValue;
+			$updated = $xpath->query('atom:updated/text()', $entry)->item(0)->nodeValue;
+	
+			if ($last_updated < $published)
+				$last_updated = $published;
+	
+			if ($last_updated < $updated)
+				$last_updated = $updated;
+		}
+	
+		// Maybe there aren't any entries. Then check if it is a valid feed
+		if ($last_updated == "") {
+			if ($xpath->query('/atom:feed')->length > 0) {
+				$last_updated = NULL_DATE;
+			}
+		}
+		q(
+			"UPDATE `gcontact` SET `updated` = '%s', `last_contact` = '%s' WHERE `nurl` = '%s'",
+			dbesc(DBM::date($last_updated)),
+			dbesc(DBM::date()),
+			dbesc(normalise_link($profile))
+		);
+	
+		if (($gcontacts[0]["generation"] == 0)) {
+			q(
+				"UPDATE `gcontact` SET `generation` = 9 WHERE `nurl` = '%s'",
+				dbesc(normalise_link($profile))
+			);
+		}
+	
+		logger("Profile ".$profile." was last updated at ".$last_updated, LOGGER_DEBUG);
+	
+		return($last_updated);
+	}
+
+	public static function doUpdate($created, $updated, $last_failure, $last_contact)
```

Comment:
This function doesn't actually do the update, please rename it to `updateNeeded()`.

Reply:
done